### PR TITLE
Feat/use long press

### PR DESCRIPTION
**Pull request by crazylxr**
[[中文版模板 / Chinese template](https://github.com/alibaba/hooks/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 💡 Background and solution

监听目标元素的长按事件。

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

File: packages/hooks/src/useLongPress/index.zh-CN.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+## API
+
+```typescript
+type Target = Element | (() => Element) | React.MutableRefObject<Element>;
+
+const isPressing = useLongPress(
+  onLongPress: (event: MouseEvent | TouchEvent) => void,
+  target: Target ,
+  delay?: number
```
Comment (awmleer) [inline packages/hooks/src/useLongPress/index.zh-CN.md]: 看上面 `index.ts` 中 `delay` 是不可缺省的，这里是写错了么

File: packages/hooks/src/useLongPress/demo/demo1.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * desc.zh-CN: 请长按按钮查看效果。
+ */
+
+import React, { useState, useRef } from 'react';
+import { useLongPress } from 'ahooks';
+
+export default () => {
+  const [counter, setCounter] = useState(0);
+  const ref = useRef<HTMLButtonElement>();
+  const isPressing = useLongPress(() => setCounter((s) => s + 1), ref, 1000);
```
Comment (awmleer) [inline packages/hooks/src/useLongPress/demo/demo1.tsx]: `isPressing` 下面好像没用到……？

File: packages/hooks/src/useLongPress/index.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import useBoolean from '../useBoolean';
+import useEventListener from '../useEventListener';
+import useLatest from '../useLatest';
+import type { BasicTarget } from '../utils/dom2';
+
+type EventType = MouseEvent | TouchEvent;
+
+function useLongPress(onLongPress: (event: EventType) => void, target: BasicTarget, delay: number) {
+  const onLongPressRef = useLatest(onLongPress);
+  const timer = useRef<ReturnType<typeof setTimeout>>();
```
Comment (awmleer) [inline packages/hooks/src/useLongPress/index.ts:21]: 这里为啥不直接用 `number`

```suggestion
  const timer = useRef<number>();
```

File: packages/hooks/src/useLongPress/index.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+  useEventListener(
+    'mouseup',
+    (e) => {
+      e.preventDefault();
+
+      endTime = Date.now();
+      if (endTime - startTime < delay) {
+        clearTimeout(timer.current!);
+      }
```
Comment (awmleer) [inline packages/hooks/src/useLongPress/index.ts]: 这里好像不用判断的，直接 clear 就行了

```suggestion
        clearTimeout(timer.current!);
```

**Review by awmleer — CHANGES_REQUESTED** [review]

File: packages/hooks/src/useLongPress/index.ts (hunk 3 above)
Comment (crazylxr) [reply packages/hooks/src/useLongPress/index.ts:21]: 看起来更直观一点，另外 node 里面的 setTimeout 好像不是 number

File: packages/hooks/src/useLongPress/index.ts (hunk 4 above)
Comment (crazylxr) [reply packages/hooks/src/useLongPress/index.ts]: 有道理，我去掉

File: packages/hooks/src/useLongPress/index.zh-CN.md (hunk 1 above)
Comment (crazylxr) [reply packages/hooks/src/useLongPress/index.zh-CN.md]: 我觉得可以缺省一下，默认 1500 ms 你觉得怎么样？

File: packages/hooks/src/useLongPress/demo/demo1.tsx (hunk 2 above)
Comment (crazylxr) [reply packages/hooks/src/useLongPress/demo/demo1.tsx]: 我加个 demo，把 isPressing 用一下。

File: packages/hooks/src/useLongPress/index.zh-CN.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+type Target = Element | (() => Element) | React.MutableRefObject<Element>;
+
+const isPressing = useLongPress(
+  onLongPress: (event: MouseEvent | TouchEvent) => void,
+  target: Target ,
+  options?: {
+    delay?: number,
+    cancelOnMovement?: boolean,
+  }
+);
```
Comment (awmleer) [inline packages/hooks/src/useLongPress/index.zh-CN.md:39]: @brickspert `target` 和 `onLongPress` 两个顺序要不要调整一下……涉及到和 `useClickAway` 保持一致
cc @ttys026

**Review by brickspert — APPROVED** [review]

**Pull request merged**